311 feature create a filtering system for student to do page

RichardM1337 opened the pull request:
### Changes
- added class filterer based on query to assignments page only (#311) 
- matched filterer with css of filter and sort

RichardM1337 committed:
added route checking

TODO:
- add dropdown that takes all courses data and sorts responses by
- connect this to route.query by class (replicating google classroom functionality)
^_^

RichardM1337 committed:
checkbox visually functional

added functinality for checlbox that works with route and replaces query
- courses are paRsed as props

assignments should now check params or emits to be properly formatted and dropdown overlap should be fixed

RichardM1337 committed:
Feature mostly added

queiries look a bit ugly and some of the todo.vue functions needed altering to compute whether a query existed.

Let me know if there are any typescript errors, logic based errors, or if you think my code is plain ugly. Thanks!

RichardM1337 committed:
fixes

now css works
other arbitary fixes
- types
- awaiting route change

no blow up plsthx

RichardM1337 committed:
Merge branch 'dev' into 311-feature-create-a-filtering-system-for-student-to-do-page

kennething commented on components/student/Todo/ByClass.vue:
```diff
@@ -0,0 +1,68 @@
+</template>
+
+<script setup lang="ts">
+const props = defineProps<{ studentCourses: StudentCourse[]; close: boolean }>();
```
1. use multiple lines for props if theres more than 1 prop
2. does studentCourses really need to be a prop? its already in store

kennething commented on components/student/Todo/ByClass.vue:
```diff
@@ -0,0 +1,68 @@
+
+async function updateQuery(courseId: number) {
+  queries.value = queries.value.includes(courseId) ? queries.value.filter((id: number) => id !== courseId) : [...queries.value, courseId];
+  await router.replace({
```
use the [`changeRouteQuery`](https://github.com/sitechtimes/regents-prep-app-frontend/blob/dev/utils/functions.ts#L40) helper function

kennething commented on components/student/Todo/ByClass.vue:
```diff
@@ -0,0 +1,68 @@
+  await router.replace({
+    query: {
+      ...route.query,
+      class: queries.value.length ? queries.value.map(String) : undefined
```
you dont need to turn everything into a string, numbers work perfectly fine for route query

kennething commented on components/student/Todo/ByClass.vue:
```diff
@@ -0,0 +1,68 @@
+        v-show="showClasses"
+        class="absolute top-10 z-[2] flex w-52 flex-col items-start justify-center gap-1 rounded-xl border-2 border-neutral-200 bg-white p-3 shadow-lg transition duration-500 dark:border-neutral-700 dark:bg-neutral-800"
+      >
+        <button v-for="course in props.studentCourses" class="du-label flex w-full cursor-pointer items-center justify-start gap-2" type="button" @click="updateQuery(course.id)">
```
dont need `props.` in template

kennething commented on components/student/Todo/ByClass.vue:
```diff
@@ -0,0 +1,68 @@
+const queries = ref<number[]>([]);
+
+async function updateQuery(courseId: number) {
+  queries.value = queries.value.includes(courseId) ? queries.value.filter((id: number) => id !== courseId) : [...queries.value, courseId];
```
dont explicitly type `id` its already implicitly typed

kennething commented on components/student/Todo/Toolbar.vue:
```diff
@@ -21,9 +22,13 @@ const emit = defineEmits<{
 const isSticky = ref(false);
 
+const route = useRoute();
+const ifTodo = ref(route.path === "/student/todo");
```
1. does this need to be a ref? you never change it
2. does this need to be a variable? you only use this value once and its in the template

kennething commented on components/student/Todo/Toolbar.vue:
```diff
@@ -1,17 +1,18 @@
 
 <script setup lang="ts">
-const props = defineProps<{ closeToolbar: boolean }>();
+const props = defineProps<{ closeToolbar: boolean; studentCourses: StudentCourse[] }>();
```
unnecessary prop, its already in store

kennething commented on pages/student/todo.vue:
```diff
@@ -43,7 +49,13 @@ const filteredAssignments = computed(() => {
+      ?.filter(filters)
+      .filter((assignment) => assignment.assignment?.name.toLowerCase().includes(search.toLowerCase()))
+      .sort(sorter)
+      .filter((assignment) => route.query.class?.includes(String(assignment.assignment.course?.id)));
```
the only difference between these 2 returns is this line. find a way to not use this entirely separate return

RichardM1337 committed:
type error and kenneth issue

fixed most of kenneth's comments
-- let me know if changeRouteQuery shouldn't take number[]

kennething commented on components/student/Todo/ByClass.vue:
```diff
@@ -0,0 +1,64 @@
+<script setup lang="ts">
+const userStore = useUserStore();
+const { studentCourses } = storeToRefs(userStore);
+const props = defineProps<{ close: boolean }>();
```
props should be at the top

kennething commented on components/student/Todo/ByClass.vue:
```diff
@@ -0,0 +1,64 @@
+const userStore = useUserStore();
+const { studentCourses } = storeToRefs(userStore);
+const props = defineProps<{ close: boolean }>();
+const showClasses = ref(false);
```
move this below route

kennething commented on components/student/Todo/ByClass.vue:
```diff
@@ -0,0 +1,64 @@
+
+async function updateQuery(courseId: number) {
+  queries.value = queries.value.includes(courseId) ? queries.value.filter((id) => id !== courseId) : [...queries.value, courseId];
+  await (queries.value.length ? changeRouteQuery({ class: queries.value }) : changeRouteQuery({ class: undefined as undefined }));
```
1. undefined as undefined?
2. ur using changeRouteQuery twice and the only thing different is the value of class, how can u ternary better?

kennething commented on pages/student/todo.vue:
```diff
@@ -43,11 +43,11 @@ const filteredAssignments = computed(() => {
     .filter((assignment) => assignment.assignment?.name.toLowerCase().includes(search.toLowerCase()))
-    .sort(sorter);
+    .sort(sorter)
+    .filter((assignment) => (route.query.class?.length ? route.query.class?.includes(String(assignment.assignment.course?.id)) : assignment));
```
this still goes through the entire assignments array and checks route.query.class for every single iteration. the assignments array could potentially be really big so this is really dumb. can we only filter if route.query.class?

kennething committed:
add type generic

RichardM1337 committed:
fixed discussions

- simplified functions
- made sure assignments aren't checked per assignment

kennething commented on pages/student/todo.vue:
```diff
@@ -43,11 +43,16 @@ const filteredAssignments = computed(() => {
     .filter((assignment) => assignment.assignment?.name.toLowerCase().includes(search.toLowerCase()))
     .sort(sorter);
+
+  if (route.query.class?.length) {
```
make this a 1 liner and stack it on top of the bottom return with no indent between

kennething reviewed (changes requested):
otherwise lgtm

RichardM1337 committed:
syntax hotfix

kennething reviewed (approved):
LGTM
